File: bun-cli/src/handlers/delete.ts
/**
 * Delete note handler using Effect.
 */

import * as p from "@clack/prompts";
import { Effect } from "effect";
import pc from "picocolors";
import type { NoteId } from "../entity/note";
import { UserCancelledError, ValidationError } from "../errors";
import { NoteService } from "../service/note-service";

/**
 * Execute delete command.
 */
export const execute = Effect.gen(function* () {
  console.log(pc.red("\n🗑️ Delete a note"));

  const service = yield* NoteService;
  const notes = yield* service.list();

  if (notes.length === 0) {
    console.log(pc.yellow("📭 No notes to delete."));
    return;
  }

  const selected = yield* Effect.tryPromise({
    try: () =>
      p.select({
        message: "Select a note to delete:",
        options: notes.map((n) => ({
          value: n.id,
          label: n.title,
          hint: n.content.slice(0, 30),
        })),
      }),
    catch: () => new ValidationError({ message: "Failed to select note" }),
  });

  if (p.isCancel(selected)) {
    return yield* Effect.fail(new UserCancelledError({ message: "Cancelled" }));
  }

  const noteId = selected as NoteId;
  const note = notes.find((n) => n.id === noteId);

  const confirmed = yield* Effect.tryPromise({
    try: () =>
      p.confirm({
        message: `Are you sure you want to delete "${note?.title}"?`,
      }),
    catch: () => new ValidationError({ message: "Failed to confirm" }),
  });

  if (p.isCancel(confirmed) || !confirmed) {
    console.log(pc.yellow("Cancelled."));
    return;
  }

  yield* service.delete(noteId);

  console.log(pc.green(`\n✅ Note "${note?.title}" deleted.`));
});

File: go-cli/internal/handler/delete.go
// Package handler contains CLI command handlers.
package handler

import (
	"fmt"

	"convex-notes-cli/internal/service"

	"github.com/AlecAivazis/survey/v2"
	"github.com/fatih/color"
)

// DeleteHandler handles the delete note command.
type DeleteHandler struct {
	service *service.NoteService
}

// NewDeleteHandler creates a new DeleteHandler.
func NewDeleteHandler(svc *service.NoteService) *DeleteHandler {
	return &DeleteHandler{service: svc}
}

// Execute runs the handler.
func (h *DeleteHandler) Execute() error {
	red := color.New(color.FgRed, color.Bold)
	green := color.New(color.FgGreen)
	yellow := color.New(color.FgYellow)

	red.Println("\n🗑️ Delete a note")

	notes, err := h.service.List()
	if err != nil {
		return fmt.Errorf("failed to list notes: %w", err)
	}

	if len(notes) == 0 {
		yellow.Println("📭 No notes to delete.")
		return nil
	}

	// Build options
	options := make([]string, len(notes))
	for i, note := range notes {
		options[i] = note.Title
	}

	var selectedIdx int
	if err := survey.AskOne(&survey.Select{
		Message: "Select a note to delete:",
		Options: options,
	}, &selectedIdx); err != nil {
		return err
	}

	selected := notes[selectedIdx]

	var confirmed bool
	if err := survey.AskOne(&survey.Confirm{
		Message: fmt.Sprintf("Are you sure you want to delete '%s'?", selected.Title),
		Default: false,
	}, &confirmed); err != nil {
		return err
	}

	if !confirmed {
		yellow.Println("Cancelled.")
		return nil
	}

	if err := h.service.Delete(selected.ID); err != nil {
		return fmt.Errorf("failed to delete note: %w", err)
	}

	green.Printf("\n✅ Note '%s' deleted.\n", selected.Title)

	return nil
}
